### ISSUE TITLE
[TF2.0] Custom training loop for keras example should provide information about tf.keras.backend.set_learning_phase()

### ISSUE THREAD

**Issue by KichangKim** [tensorflow/tensorflow#28028]
**System information**
- TensorFlow version: tensorflow-gpu 2.0.0-alpha0
- Doc Link: https://www.tensorflow.org/alpha/guide/keras/training_and_evaluation#using_the_gradienttape_a_first_end-to-end_example

**Describe the documentation issue**
That samples did not use tf.keras.backend.set_learning_phase() for manual training loop. So if the model has BatchNormalization() or other layers which depend on training state, it will be not trained correctly. There is no description about tf.keras.backend.set_learning_phase(), even in the page of BatchNormalization(). (https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/layers/BatchNormalization)

**Comment by nartes** [comment]
Are there any examples on a custom training loop with BatchNormalization?
I've managed to set training_phase, although the model doesn't train the same way as via keras default training loop.

**Comment by nartes** [comment]
1. Keras has different side operations which depends on whether train, test or predict phase is executed.
- https://github.com/keras-team/keras/blob/master/keras/engine/training.py#L497
- https://github.com/keras-team/keras/blob/master/keras/engine/training.py#L524
- https://github.com/keras-team/keras/blob/master/keras/engine/training.py#L542
2. In general a layer has unconditional and condition on input updates and losses. Updates can be that of running mean average in batch normalization.
https://github.com/keras-team/keras/blob/master/keras/layers/normalization.py#L197
Layer losses may include regularization.
https://github.com/keras-team/keras/blob/master/keras/engine/base_layer.py#L246
3. Upon first call to K.learning_phase() a placeholder for the phase state is created with default value False. To train the model feed dict should contain {K.learning_phase(): True}.
https://github.com/keras-team/keras/blob/master/keras/backend/tensorflow_backend.py#L123
https://github.com/keras-team/keras/blob/master/keras/backend/tensorflow_backend.py#L3345
https://github.com/keras-team/keras/blob/master/keras/layers/normalization.py#L206

**Comment by KichangKim** [comment]
I simply called set training_phase before creating the model, and train the model with same method in TF 2.0 tutorial. Without calling it, the model does not trained well.

**Comment by KichangKim** [comment]
I found that other document describes "training=True" flag for manual training loop:

https://www.tensorflow.org/alpha/guide/migration_guide#customize_the_training_step

But this does not:

https://www.tensorflow.org/alpha/guide/keras/training_and_evaluation#using_the_gradienttape_a_first_end-to-end_example

Also, second document does not use tf.function for its training step, so extremely slow than first sample. Additionally, GradientTape version manual trainig loop requires more memory than train_on_batch version. I can't figure out that this issue is intended behaviour or bug. (Maybe related to #28110)

**Comment by robieta** [comment]
With respect to the initial issue, I don't thing a discussion of the learning phase is appropriate, because the example doesn't include any layers that depend on it. Writing a custom loop that uses batch norm is something of an advanced use case both because of the change in behavior between training and inference and the fact that its updates are not gradient updates. All of that complexity would distract from the goal of a simple first example.

With respect to the actual use of the learning phase, I would recommend directly using the `training` argument of the model rather than `set_learning_phase`. Passing what you want directly into a function (a model is a function, after all) is more idiomatic TF 2.0 than setting a global persistent state. (Or `keras.backend.learning_phase_scope` can split the difference) I do, however, agree that the learning phase is not always a straightforward concept and it can be easy to get tripped up on layers whose behavior depend on it. @dynamicwebpaige @fchollet Thoughts on adding a learning phase example to the documentation queue?

Lastly, with respect to performance what you describe is expected. Eager performance is expected to be slower than graph in most cases, and higher memory consumption is also expected. In graph mode the runtime is free to free memory during backprop, but in eager it must simultaneously hold all of the activations and all of the gradients in memory. As for acceleration on with tf.function: there are definitely guides on that, but it has some quirks (function tracing and only once execution, trace caching, etc.) that again makes me think it probably shouldn't be in the very first run loop example.

**Comment by olesalscheider** [comment]
I agree that it would be nice to have an example with batch normalization.

I also stumbled about this when I converted my code to TF 2: In graph mode you had to run the collected update ops from your model. In eager mode all updates were done in place and the call to updates() just did nothing. But once I wrapped my training step in a tf.function I ended up getting an error and it was really not clear to me if I had to run the updates() inside a tf.function or not. I think it might be helpful to clarify that.

**Comment by KichangKim** [comment]
> Lastly, with respect to performance what you describe is expected. Eager performance is expected to be slower than graph in most cases, and higher memory consumption is also expected. In graph mode the runtime is free to free memory during backprop, but in eager it must simultaneously hold all of the activations and all of the gradients in memory. As for acceleration on with tf.function: there are definitely guides on that, but it has some quirks (function tracing and only once execution, trace caching, etc.) that again makes me think it probably shouldn't be in the very first run loop example.

@robieta Thanks for clarification. I think that this also should be clearly described in documents for new users. Current training sample (Eager without tf.function) is extremely slow as you described, so it makes user misunderstand that TF 2.0 is slow (like me).

I know that TF 2.0 is in the early stage, but these issue makes me that hard to investigate whether an issue is intended behaviour or bug without documentations.

**Comment by edoardogiacomello** [comment]
With regard to the tf.keras.BatchNormalization layer, the documentation specifies that the second input parameter is a python boolean indicating whether the model is used for training or not.
How does this relate to the trainable parameter of model? If one left the layer flag unspecified, would the Model parameter be used instead? 
What about if one wants to alternate a train and test step, is it sufficient to edit the model variable or should it be rebuilt? 

I think that the documentation is lacking this kind of information, which is quite crucial for successfully using recent models.

**Comment by KichangKim** [comment]
Additionally, when I using Model.train_on_batch(), there is no way to pass Training=True flag, so, does TF 2.0 automatically set True when calling train_on_batch() or should I set set_learning_phase() manually?

Current documentation is too useless for that situation.

**Comment by saberkun** [comment]
I think this is a hidden thing in tf.keras now.
With keras model.fit() api, all these are hidden. @robieta documentation is necessary.
However, if you want to directly use keras layers and models, there is a training field inside keras model. For evaluation, you shall set it as false. https://github.com/tensorflow/models/blob/master/official/bert/model_training_utils.py#L249

Correct me if I am wrong....
Thx

**Comment by celinew1221** [comment]
The migration guide actually have it for TF2.0: https://www.tensorflow.org/beta/guide/migration_guide. But it wasn't mentioned in keyword. So it's sort of hidden here.

**Comment by celinew1221** [comment]
> **System information**
> 
> * TensorFlow version: tensorflow-gpu 2.0.0-alpha0
> * Doc Link: https://www.tensorflow.org/alpha/guide/keras/training_and_evaluation#using_the_gradienttape_a_first_end-to-end_example
> 
> **Describe the documentation issue**
> That samples did not use tf.keras.backend.set_learning_phase() for manual training loop. So if the model has BatchNormalization() or other layers which depend on training state, it will be not trained correctly. There is no description about tf.keras.backend.set_learning_phase(), even in the page of BatchNormalization(). (https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/layers/BatchNormalization)

Additionally, using `training=False` or `training=True` doesn't set learning phase in keras. So when I print `keras.backend.learning_phase()`, it will also say 0 (test phase). This is confusing.

**Comment by nartes** [comment]
> Additionally, using `training=False` or `training=True` doesn't set learning phase in keras. So when I print `keras.backend.learning_phase()`, it will also say 0 (test phase) even if I use the training argument. This is confusing.
It doesn't because training variable directly affect which computing branch is return at graph construction time inside a python interpreter.
https://github.com/keras-team/keras/blob/master/keras/backend/tensorflow_backend.py#L3363

**Comment by rwl93** [comment]
> Are there any examples on a custom training loop with BatchNormalization?
> I've managed to set training_phase, although the model doesn't train the same way as via keras
> default training loop.

You can find an example of the `training` variable usage for batch norm layers using the functional API and custom training loops in the Tensorflow resnet50 implementation. 

[Here](https://github.com/tensorflow/models/blob/e38d570ef03c16c26d037b23fc2f35ecd4f87032/official/resnet/ctl/ctl_imagenet_main.py#L184) is the model call with the `training=True` in the custom training loop.

Note that the model definition is using the functional API and does **not** specify the batch norm layers with the `training` parameter. [see here](https://github.com/tensorflow/models/blob/e38d570ef03c16c26d037b23fc2f35ecd4f87032/official/resnet/keras/resnet_model.py#L202)

When the custom training loop calls the model with the `training` parameter, the [`call` method in tensorflow/python/keras/engine/network.py](https://github.com/tensorflow/tensorflow/blob/7c2a7d4684b55f6f869bd814d67f2d2d4629fd55/tensorflow/python/keras/engine/network.py#L675) calls `_run_internal_graph` which resets the `training` parameter for each layer that has `'training'` in its argspec. [see here](https://github.com/tensorflow/tensorflow/blob/7c2a7d4684b55f6f869bd814d67f2d2d4629fd55/tensorflow/python/keras/engine/network.py#L828-L830)

**Comment by Bakaizhamgyrchiev** [comment]
What about models from keras.applications?
I would to like to use keras.Model from keras.applications in my custom training loop.

It is not clear, because these models create like tf.keras.Model(inputs, outputs) and there is not any information about call function parameters.

**Comment by martin-gorner** [comment]
@rwl93 , I cannot find any special handling of batch norm for custom training loops in the code you linked
@robieta , you mention that custom training loops with batch norm are an advanced topic

So which is it ? When I try, my model is not training to the same accuracy as with model.fit(). But then I might be hitting another issue.

Is there something special to do when using tf.keras.layers.BatchNormalization with a custom training loop ?

**Comment by robieta** [comment]
I put together a quick example of batch norm and CTL [gist](https://colab.sandbox.google.com/gist/robieta/044ec2031399e1dae5e42fc7c8bd48d3/batch_norm_in_ctl.ipynb). The main thing is to pass the training arg when calling the model. (I was lazy and just used model.evaluate to assess the CTL) Interestingly that also seems to apply the updates to the moving statistics. For whatever reason I thought those had to be handled separately (handling them correctly was a major stumbling block of 1.x CTL's), but it seems that either the layer or automatic control dependencies are picking them up automagically. So I guess no so advanced a use case as I initially thought.

**Comment by saberkun** [comment]
I also noticed the problem months ago. I remember none of our examples pass
training arg. However, we should pass it explicitly in CTL and all official
models have that now.

**Comment by martin-gorner** [comment]
Thank you for putting this example together. It is really nice. You say the main thing is to call the model with model(data, training=True) during training. After reading your code, it seems to me that's the *only* thing you are doing differently. Correct ?

**Comment by robieta** [comment]
That is correct.

**Comment by Bakaizhamgyrchiev** [comment]
Recommend to look at this method
https://github.com/tensorflow/tensorflow/blob/99dda8ef941eac413e196b7bd6a7ba58b553b4ca/tensorflow/python/keras/engine/base_layer.py#L627

it calls before every specific layer calls

firstly training flag is searched in kwargs, and if not training= backend.learning_phase

**Comment by KichangKim** [comment]
> I put together a quick example of batch norm and CTL [gist](https://colab.sandbox.google.com/gist/robieta/044ec2031399e1dae5e42fc7c8bd48d3/batch_norm_in_ctl.ipynb). The main thing is to pass the training arg when calling the model. (I was lazy and just used model.evaluate to assess the CTL) Interestingly that also seems to apply the updates to the moving statistics. For whatever reason I thought those had to be handled separately (handling them correctly was a major stumbling block of 1.x CTL's), but it seems that either the layer or automatic control dependencies are picking them up automagically. So I guess no so advanced a use case as I initially thought.

@robieta  Thanks for sample code. It clarifies how to train BatchNormalization layer by using GradientTape. But how about train_on_batch() and predict_on_batch() API? it doesn't have training parameter. In my understand, keras API (train, predict, evaluate and so on) handles training status internally (so it will be True while I running train_xx method) and I should not call set_learning_phase() manually. Is this correct?

**Comment by robieta** [comment]
That is correct: the `Model.*_on_batch` methods set the learning phase for you.

**Comment by alanchiao** [comment]
With regards to the original issue, I made [this commit](https://github.com/tensorflow/docs/commit/7ad5e9598e66efeef8b3ed9fa53fa5ed023ddbed) that gives users a heads up about the training=True/False for custom training loops across the tutorials on tensorflow.org where possible (some examples have single layers that don't have or need a training=True/False and trying to pass the training param would throw an error).
